### PR TITLE
Cask loader: add missing canonicalisation

### PR DESCRIPTION
**Pull request by claui**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally? **A couple of tests fail** on my local machine but all of them seem to be unrelated. See #7417.

-----

This PR fixes a bug I missed in PR #7417.
Also, a flaw in our test helper masked the bug so it went past the radar.

This PR has four steps:

1. Change an interface for testability.

2. Add a test for the PR #7417 bug.  
   That test is supposed to fail but actually passes due to a flaw in our test helper.  
   In other words: this commit reveals a test helper bug.

3. Fix the test helper bug. Now the test fails, exposing the PR #7417 bug.

4. Fix the PR #7417 bug.

For details to one of the steps, see the detailed commit message of the individual step.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
Thanks @claui!
